### PR TITLE
Basic working version of uploading and saving states

### PR DESCRIPTION
**Pull request by kennedyry**
ℹ️ Issue
Closes clickup ticket link

📝 Description
Briefly list the changes made to the code:
* Setup backend schemas to work with sending update requests when timesheets are now modified. 
* Updated all relevant files to support update functionality of timesheet modifications. 
* Updated tab functionality to use a state tracking the selected tab option removing the need for a blank timesheet to track this. 

✔️ Testing
What steps did you take to verify your changes work? These should be clear enough for someone to be able to clone the branch and follow the steps themselves.
* Tested various modifications to the timesheet and ensured that they were properly modified. 

Provide screenshots of any new components, styling changes, or pages.

🏕️ (Optional) Future Work / Notes
Did you notice anything ugly during the course of this ticket? Any bugs, design challenges, or unexpected behavior? Write it down so we can clean it up in a future ticket!

### CODE REVIEW

File: src/components/TimeCardPage/TimeSheet.tsx [hunk 1]
```diff
@@ -51,7 +51,7 @@ const createEmptyTable = (startDate, company) => {
     //TODO's: Pull UserID automatically
     return {
         UserID: "77566d69-3b61-452a-afe8-73dcda96f876",
-        TimesheetID: uuidv4(), 
+        TimesheetID: Math.round(Math.random() * 1000000000),
```
Comment (circlegov) [inline src/components/TimeCardPage/TimeSheet.tsx:54]: Why this change?

**Review by circlegov — APPROVED** [review]
Looks great, just a question abt the uuid change

File: src/components/TimeCardPage/TimeSheet.tsx (hunk 1 above)
Comment (kennedyry) [reply src/components/TimeCardPage/TimeSheet.tsx:54]: Oh the type was a number whereas uuidv4 is a string so I just created very large number as a workaround for the moment I should probably actually fix this

File: src/components/TimeCardPage/TimeTableRow.tsx [hunk 2]
```diff
@@ -12,29 +12,50 @@ import {TIMEZONE} from 'src/constants';
...
+        
         const newFields = {
             ...fields, 
             [key]: value 
         }
+
         setFields(newFields); 
         props.onRowChange(newFields); 
+        //Send a request to update the db on this item being changed 
+        ApiClient.updateTimesheet(updateSchemas.TimesheetUpdateRequest.parse({
```
Comment (izzyconner) [inline src/components/TimeCardPage/TimeTableRow.tsx:49]: Is this going to be called every time a field is changed, or only when the user clicks 'save'? either way is probably fine for now, but it might be a little excessive if it's every time something is changed. if that's the case, maybe just leave a TODO for adding in a timed auto-save instead?

File: src/schemas/backend/UpdateTimesheet.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    INSERT = "INSERT", 
+    UPDATE = "UPDATE", 
+    DELETE = "DELETE", 
+    STATUS_CHANGE = "STATUS_CHANGE", 
+    CREATE_TIMESHEET = "CREATE_TIMESHEET"
+} 
+
+
+
+export const enum TimesheetListItems {
```
Comment (izzyconner) [inline src/schemas/backend/UpdateTimesheet.ts:25]: Could you add in some quick comments for the enums here? I think they're hopefully self-explanatory, but will be nice to have it documented for new devs. Maybe also just give a couple sentence summary at the top of this file explaining what the whole idea behind having these CRUD-like operations is?

File: src/schemas/backend/UpdateTimesheet.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    Type: availableListTypes, 
+    Id: z.string(), 
+    Attribute: z.string(), 
+    Data: z.any() 
+})
+export type UpdateRequest = z.infer<typeof UpdateRequest>
+
+
+// The main request body that is used to determine what we should be updating in a request 
+export const TimesheetUpdateRequest = z.object({
```
Comment (izzyconner) [inline src/schemas/backend/UpdateTimesheet.ts:61]: super clean - I love this

File: src/schemas/backend/UpdateTimesheet.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+}) 
+export type InsertRequest = z.infer<typeof InsertRequest> 
+/*
+    Schema for updating an item from the three possible list of items in the timesheet 
+    Type: The field of the timesheet we are updating from the three supported 
+    Id: the id of the entry we are updating - correlates to that row / entry in the list of items 
+    Attribute: The specific attribute of the object we are updating 
+    Data: The payload we are updating this attribute to be - can be a wide range of things currently 
+*/
+export const UpdateRequest = z.object({
```
Comment (izzyconner) [inline src/schemas/backend/UpdateTimesheet.ts:51]: great comment explaining the schema

File: src/schemas/backend/UpdateTimesheet.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    TABLEDATA = "TABLEDATA", 
+    SCHEDULEDATA = "SCHEDULEDATA", 
+    WEEKNOTES = "WEEKNOTES"
+}
+
+const availableListTypes = z.enum([TimesheetListItems.TABLEDATA, TimesheetListItems.SCHEDULEDATA, TimesheetListItems.WEEKNOTES])
+
+export const DeleteRequest = z.object({
+    Type: availableListTypes, 
+    Id: z.string()
```
Comment (izzyconner) [inline src/schemas/backend/UpdateTimesheet.ts:35]: Could you leave a tiny tiny comment explaining what the 'id' is here? I know that it's pretty much the same as what you explain for the `UpdateRequest`, but might be helpful to explain it for all request types in case people get confused.

**Review by izzyconner — COMMENTED** [review]
looks really really good! I think that there could just be a little more comments/documentation for the actual `UpdateTimesheet` file since it's a major overhaul of how we are doing timesheet requests. other than that, good to go!

File: src/components/TimeCardPage/TimeTableRow.tsx (hunk 2 above)
Comment (kennedyry) [reply src/components/TimeCardPage/TimeTableRow.tsx:49]: At some point we had discussed it and I believe the consensus was to call it everytime mainly since our current setup handles a single change at a time - uploading at saves / an auto-save time would require us to track these changes and then upload a list of them at a time or call it several times at that point removing from the queue of changes - should probably discuss this in a meeting / plan a new ticket for this if we want to change how this works as at the moment the endpoint just accepts a single update at a time.

File: src/schemas/backend/UpdateTimesheet.ts (hunk 3 above)
Comment (kennedyry) [reply src/schemas/backend/UpdateTimesheet.ts:25]: Yeah makes sense will update it in the backend / leave comment to check there instead as its shared code / will be pointing to backend at some point with monorepo.

File: src/schemas/backend/UpdateTimesheet.ts (hunk 6 above)
Comment (kennedyry) [reply src/schemas/backend/UpdateTimesheet.ts:35]: updated on backend portion of repo.

**Comment by kennedyry** [comment]
@circlegov Since we are trying to generate timesheets on the backend I was trying to refactor out the createEmptyTimesheet method so that the frontend was never creating them which required me to tweak a portion of the code you used for tabs where I just basically altered it so that we were no longer using a blank timesheet to signify the aggregated tabs if this creates any problems lmk
